dev/src/document-reader.ts

```diff
@@ -0,0 +1,193 @@
+/*!
+ * Copyright 2021 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {DocumentSnapshot, DocumentSnapshotBuilder} from './document';
+import {DocumentReference} from './reference';
+import {FieldPath} from './path';
+import {isPermanentRpcError} from './util';
+import {google} from '../protos/firestore_v1_proto_api';
+import {logger} from './logger';
+import {Firestore} from './index';
+
+import api = google.firestore.v1;
+
+/**
+ * A wrapper around BatchGetDocumentsRequest that retries request upon stream
+ * failure and returns ordered results.
+ *
+ * @private
+ */
+export class DocumentReader<T> {
+  /** An optional field mask to apply to this read. */
+  fieldMask?: FieldPath[];
+  /** An optional transaction ID to use for this read. */
+  transactionId?: Uint8Array;
+
+  private remainingDocuments = new Set<string>();
+  private retrievedDocuments = new Map<string, DocumentSnapshot>();
+
+  /**
+   * Internal method to retrieve multiple documents from Firestore, optionally
+   * as part of a transaction.
+   *
+   * @param firestore The Firestore instance to use.
+   * @param allDocuments The documents to receive.
+   * @returns A Promise that contains an array with the resulting documents.
+   */
+  constructor(
+    private firestore: Firestore,
+    private allDocuments: Array<DocumentReference<T>>
+  ) {
+    for (const docRef of this.allDocuments) {
+      this.remainingDocuments.add(docRef.formattedName);
+    }
+  }
+
+  /**
+   * Invokes the BatchGetDocuments RPC and returns the results.
+   *
+   * @param requestTag A unique client-assigned identifier for this request.
+   */
+  async get(requestTag: string): Promise<Array<DocumentSnapshot<T>>> {
```

Comment:
If get() is only meant to be run once, the method should throw an error if it is run again

Reply:
It is only meant to be run once, but it should work if run multiple times. It will return the same result without doing any work.